File: src/lib/platform/engine.ts
import { COMPETENCIES, IGOT_COURSES } from "./data";
import type { Course, GapRow, GapSeverity } from "./types";

export function classifyGap(gap: number): GapSeverity {
  if (gap <= 0) return "No Gap";
  if (gap === 1) return "Low";
  if (gap === 2) return "Medium";
  if (gap === 3) return "High";
  return "Critical";
}

export const SEVERITY_COLORS: Record<GapSeverity, string> = {
  "No Gap": "hsl(152 55% 40%)",
  Low: "hsl(196 60% 45%)",
  Medium: "hsl(38 92% 50%)",
  High: "hsl(20 90% 52%)",
  Critical: "hsl(0 72% 50%)",
};

export function buildGapRows(levels: Record<string, number>): GapRow[] {
  return COMPETENCIES.map((competency) => {
    const current = levels[competency.id] ?? 0;
    const gap = Math.max(0, competency.required - current);
    return {
      competency,
      current,
      required: competency.required,
      gap,
      severity: classifyGap(gap),
    };
  });
}

export const RECOMMENDATION_WEIGHTS = [
  { key: "skillGap", label: "Skill Gap", weight: 0.35 },
  { key: "roleRelevance", label: "Role Relevance", weight: 0.25 },
  { key: "learningHistory", label: "Learning History", weight: 0.15 },
  { key: "difficultyMatch", label: "Difficulty Match", weight: 0.1 },
  { key: "departmentPriority", label: "Department Priority", weight: 0.1 },
  { key: "careerRelevance", label: "Career Relevance", weight: 0.05 },
] as const;

export interface ScoredCourse {
  course: Course;
  score: number;
  components: Record<string, number>;
  targetedGaps: GapRow[];
}

export function recommendCourses(
  levels: Record<string, number>,
  completedCourseIds: string[],
  enrolledCourseIds: string[],
): ScoredCourse[] {
  const rows = buildGapRows(levels);
  const rowById = new Map(rows.map((r) => [r.competency.id, r]));

  const scored = IGOT_COURSES.map<ScoredCourse>((course) => {
    const targetedGaps = course.competencyIds
      .map((id) => rowById.get(id))
      .filter((r): r is GapRow => Boolean(r));

    const maxGap = targetedGaps.reduce((m, r) => Math.max(m, r.gap), 0);
    const skillGap = Math.min(1, maxGap / 4);

    // Learning history: courses that build on already-started areas score higher,
    // and already-completed courses are pushed down.
    const touchedAreas = targetedGaps.filter((r) => r.current > 0).length;
    let learningHistory = targetedGaps.length
      ? touchedAreas / targetedGaps.length
      : 0.5;
    if (completedCourseIds.includes(course.id)) learningHistory = 0;
    if (enrolledCourseIds.includes(course.id)) learningHistory *= 0.6;

    const avgCurrent = targetedGaps.length
      ? targetedGaps.reduce((s, r) => s + r.current, 0) / targetedGaps.length
      : 0;
    const difficultyMatch = Math.max(
      0,
      1 - Math.abs(course.level - (avgCurrent + 1)) / 4,
    );

    const components: Record<string, number> = {
      skillGap,
      roleRelevance: course.roleRelevance,
      learningHistory,
      difficultyMatch,
      departmentPriority: course.departmentPriority,
      careerRelevance: course.careerRelevance,
    };

    const score = RECOMMENDATION_WEIGHTS.reduce(
      (sum, w) => sum + (components[w.key] ?? 0) * w.weight,
      0,
    );

    return { course, score: Math.round(score * 1000) / 10, components, targetedGaps };
  });

  return scored.sort((a, b) => b.score - a.score);
}

/** Score an assessment and translate percentage-correct into a 1-5 level. */
export function scoreToLevel(correct: number, total: number): number {
  if (total === 0) return 1;
  const pct = correct / total;
  if (pct >= 0.9) return 5;
  if (pct >= 0.75) return 4;
  if (pct >= 0.55) return 3;
  if (pct >= 0.35) return 2;
  return 1;
}

export function overallCompetency(levels: Record<string, number>): number {
  const values = COMPETENCIES.map((c) => levels[c.id] ?? 0);
  const avg = values.reduce((a, b) => a + b, 0) / (values.length || 1);
  return Math.round(avg * 10) / 10;
}

File: stack-connect.worktrees/execute-program-request/backend/app/recommendation/engine.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, Iterable, List

WEIGHTS: Dict[str, float] = {
    "skill_gap": 0.35,
    "role_relevance": 0.25,
    "learning_history": 0.15,
    "difficulty_match": 0.10,
    "department_priority": 0.10,
    "career_relevance": 0.05,
}

assert abs(sum(WEIGHTS.values()) - 1.0) < 1e-9, "recommendation weights must sum to 1"


@dataclass
class ScoredCourse:
    course: Dict[str, Any]
    score: float
    components: Dict[str, float] = field(default_factory=dict)
    targeted_gaps: List[Dict[str, Any]] = field(default_factory=list)


def _clamp(value: float, low: float = 0.0, high: float = 1.0) -> float:
    return max(low, min(high, value))


def recommend(
    gap_rows: Iterable[Dict[str, Any]],
    courses: Iterable[Dict[str, Any]],
    completed_course_ids: Iterable[str] = (),
    enrolled_course_ids: Iterable[str] = (),
    limit: int | None = None,
) -> List[ScoredCourse]:
    """Rank courses for one officer.

    gap_rows: [{competency_id, current, required, gap, severity}, ...]
    courses:  catalogue rows with competency_ids and relevance metadata
    """
    rows = {row["competency_id"]: row for row in gap_rows}
    completed = set(completed_course_ids)
    enrolled = set(enrolled_course_ids)

    results: List[ScoredCourse] = []
    for course in courses:
        targeted = [rows[cid] for cid in course["competency_ids"] if cid in rows]

        max_gap = max((row["gap"] for row in targeted), default=0)
        skill_gap = _clamp(max_gap / 4.0)

        touched = sum(1 for row in targeted if row["current"] > 0)
        learning_history = (touched / len(targeted)) if targeted else 0.5
        if course["id"] in completed:
            learning_history = 0.0
        elif course["id"] in enrolled:
            learning_history *= 0.6

        avg_current = (
            sum(row["current"] for row in targeted) / len(targeted) if targeted else 0.0
        )
        difficulty_match = _clamp(1 - abs(course["level"] - (avg_current + 1)) / 4.0)

        components = {
            "skill_gap": skill_gap,
            "role_relevance": float(course["role_relevance"]),
            "learning_history": learning_history,
            "difficulty_match": difficulty_match,
            "department_priority": float(course["department_priority"]),
            "career_relevance": float(course["career_relevance"]),
        }

        score = sum(components[key] * weight for key, weight in WEIGHTS.items())
        results.append(
            ScoredCourse(
                course=course,
                score=round(score * 100, 1),
                components=components,
                targeted_gaps=[row for row in targeted if row["gap"] > 0],
            )
        )

    results.sort(key=lambda item: item.score, reverse=True)
    return results[:limit] if limit else results
